add an option for masking out islands in ionosphere computation

CunrenLiang opened the pull request:
Islands are not connected with lands, and thus lead to phase unwrapping errors. This might be a big issue in ionosphere computation. This pull request add an option for masking out the islands in topsApp.py.

A comparison of ionosphere computation results before filtering with and without using this option is shown in the attached figure.

![Screen Shot 2021-06-25 at 7 13 05 AM](https://user-images.githubusercontent.com/56097947/123354103-839ed500-d595-11eb-9986-a7eb9eaa00f0.jpg)

CunrenLiang committed:
add an option for masking out islands in ionosphere computation

hfattahi commented on applications/topsApp.py:
```diff
@@ -519,6 +519,14 @@
     mandatory=False,
     doc='')
 
+ION_MASKED_AREAS = Application.Parameter('ION_maskedAreas',
```
You added a very clear explanation of this added parameter to topsApp.xml. I would actually add that comment here.

CunrenLiang replied:
OK, I also copied the comments here.

hfattahi reviewed (approved):
Thanks @CunrenLiang . LGTM . I added one minor comment in case if you want to address. Even without that, this PR can be merged in my opinion.

CunrenLiang committed:
Update topsApp.py

add some comments for new parameter "areas masked out in ionospheric phase estimation"

hfattahi commented:
@yunjunz, @piyushrpt  would you please take a look at this PR.

yunjunz commented:
Thank you @CunrenLiang for the useful feature. I set up a mask-out area based on the `merged/filt_topophase.unw` file and got the following error during the testing, due to the inappropriate line/col number I guess? 

There are several numbers of looks parameters in the topsApp.xml, which can be confusing. Could you elaborate more on which file (and num of looks parameter names) we should use (refer to) as the reference to specify the line/column number in the list?

```
Writing geotrans to VRT for ion/lower/merged/topophase.flat
Multilooking ion/upper/merged/40rlks_10alks/topophase.flat ...
GDAL open (R): ion/upper/merged/40rlks_10alks/topophase.flat.vrt
API open (WR): ion/upper/merged/topophase.flat
GDAL close: ion/upper/merged/40rlks_10alks/topophase.flat.vrt
API close:  ion/upper/merged/topophase.flat
API open (R): ion/upper/merged/topophase.flat
API close:  ion/upper/merged/topophase.flat
Writing geotrans to VRT for ion/upper/merged/topophase.flat
computing ionosphere
Traceback (most recent call last):
  File "/home/zyunjun/tools/isce2/install_ion/bin/topsApp.py", line 1077, in <module>
    insar.run()
  File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/iscesys/Component/Application.py", line 142, in run
    exitStatus = self._processSteps()
  File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/iscesys/Component/Application.py", line 405, in _processSteps
    result = func(*pargs, **kwargs)
  File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/Factories.py", line 40, in __call__
    return self.method(self.other, *args, **kwargs)
  File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/runIon.py", line 2656, in runIon
    rawion(self, ionParam)
  File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/runIon.py", line 2631, in rawion
    ionosphere(self, ionParam)
  File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/runIon.py", line 1290, in ionosphere
    maskedAreas = reformatMaskedAreas(ionParam.maskedAreas, length, width)
  File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/Alos2Proc/runIonFilt.py", line 753, in reformatMaskedAreas
    raise Exception('area {} masked out in ionospheric phase estimation not correct'.format(i+1))
Exception: area 1 masked out in ionospheric phase estimation not correct
```

CunrenLiang commented:
> Thank you @CunrenLiang for the useful feature. I set up a mask-out area based on the `merged/filt_topophase.unw` file and got the following error during the testing, due to the inappropriate line/col number I guess?
> 
> There are several numbers of looks parameters in the topsApp.xml, which can be confusing. Could you elaborate more on which file (and num of looks parameter names) we should use (refer to) as the reference to specify the line/column number in the list?
> 
> ```
> Writing geotrans to VRT for ion/lower/merged/topophase.flat
> Multilooking ion/upper/merged/40rlks_10alks/topophase.flat ...
> GDAL open (R): ion/upper/merged/40rlks_10alks/topophase.flat.vrt
> API open (WR): ion/upper/merged/topophase.flat
> GDAL close: ion/upper/merged/40rlks_10alks/topophase.flat.vrt
> API close:  ion/upper/merged/topophase.flat
> API open (R): ion/upper/merged/topophase.flat
> API close:  ion/upper/merged/topophase.flat
> Writing geotrans to VRT for ion/upper/merged/topophase.flat
> computing ionosphere
> Traceback (most recent call last):
>   File "/home/zyunjun/tools/isce2/install_ion/bin/topsApp.py", line 1077, in <module>
>     insar.run()
>   File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/iscesys/Component/Application.py", line 142, in run
>     exitStatus = self._processSteps()
>   File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/iscesys/Component/Application.py", line 405, in _processSteps
>     result = func(*pargs, **kwargs)
>   File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/Factories.py", line 40, in __call__
>     return self.method(self.other, *args, **kwargs)
>   File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/runIon.py", line 2656, in runIon
>     rawion(self, ionParam)
>   File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/runIon.py", line 2631, in rawion
>     ionosphere(self, ionParam)
>   File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/TopsProc/runIon.py", line 1290, in ionosphere
>     maskedAreas = reformatMaskedAreas(ionParam.maskedAreas, length, width)
>   File "/home/zyunjun/tools/isce2/install_ion/packages/isce/components/isceobj/Alos2Proc/runIonFilt.py", line 753, in reformatMaskedAreas
>     raise Exception('area {} masked out in ionospheric phase estimation not correct'.format(i+1))
> Exception: area 1 masked out in ionospheric phase estimation not correct
> ```

Hi @yunjunz, in the updated input xml file, there are comments about reference image that can be used to set line/column numbers of masked-out areas:
https://github.com/CunrenLiang/isce2/blob/main-crl/examples/input_files/topsApp.xml#L204

yunjunz reviewed (approved):
Thank you @CunrenLiang, I should have read it more carefully!

Testing with one pair of Sentinel-1 looks good to me without issue.
